src/common/subprocess/subprocess.h

```diff
@@ -0,0 +1,370 @@
+                                                        const char *local_uri,
+                                                        uint32_t rank);
+
+/*  Stop a subprocess server / cleanup flux_subprocess_server_t */
+void flux_subprocess_server_stop (flux_subprocess_server_t *s);
+
+/* Terminate all subprocesses started by a sender id */
+int flux_subprocess_server_terminate_by_uuid (flux_subprocess_server_t *s,
+                                              const char *id);
+
+/*  General output callback that will send output from the subprocess
+ *  to stdout or stderr.  Set to `on_stdout` and/or `on_stderr` in
+ *  flux_subprocess_ops_t.  Can also be used for 'on_channel_out'
+ *  callback, sending all output to stdout.
+ */
+void flux_subprocess_output (flux_subprocess_t *p, const char *stream);
+
+/*
+ *  Commands:
+ */
+
+/*
+ *  Create a cmd object, from which subprocesses can be created
+ */
+flux_cmd_t * flux_cmd_create (int argc, char *argv[], char **env);
+
+/*
+ *  Create a copy of a cmd object.
+ */
+flux_cmd_t * flux_cmd_copy (const flux_cmd_t *cmd);
+
+/*
+ *  Destroy and free command object `cmd`
+ */
+void flux_cmd_destroy (flux_cmd_t *cmd);
+
+/*
+ *  Append formatted string to argv of `cmd`.
+ */
+int flux_cmd_argv_append (flux_cmd_t *cmd, const char *fmt, ...);
+
+/*
+ *  Return the current argument count for `cmd`.
+ */
+int flux_cmd_argc (const flux_cmd_t *cmd);
+
+/*
+ *  Return the current argument at index n (range 0 to argc - 1)
+ */
+const char *flux_cmd_arg (const flux_cmd_t *cmd, int n);
+
+/*
+ *  Set a single environment variable (name) to formatted string `fmt`.
+ *   If `overwrite` is non-zero then overwrite any existing setting for `name`.
+ */
+int flux_cmd_setenvf (flux_cmd_t *cmd, int overwrite,
+		      const char *name, const char *fmt, ...);
+
+/*
+ *  Unset environment variable `name` in the command object `cmd`.
+ */
+void flux_cmd_unsetenv (flux_cmd_t *cmd, const char *name);
+
+/*
+ *  Return current value for environment variable `name` as set in
+ *   command object `cmd`. If environment variable is not set then NULL
+ *   is returned.
+ */
+const char *flux_cmd_getenv (const flux_cmd_t *cmd, const char *name);
+
+/*
+ *  Set/get the working directory for the command `cmd`.
+ */
+int flux_cmd_setcwd (flux_cmd_t *cmd, const char *cwd);
+const char *flux_cmd_getcwd (const flux_cmd_t *cmd);
+
+/*
+ *  Request a channel for communication between process and caller.
+ *   Callers can write to the subproces via flux_subprocess_write()
+ *   and read from it via flux_subprocess_read(), which is typically
+ *   called from a callback set in 'on_channel_out'.
+ *
+ *  The `name` argument is also used as the name of an environment variable
+ *   in the subprocess environment that is set to the file descriptor number
+ *   of the process side of the socketpair. E.g. name = "FLUX_PMI" would
+ *   result in the environment variable "FLUX_PMI_FD=N" set in the process
+ *   environment.
```

Comment:
Ahhh, I didn't think about that.  I assume the ```_FD``` style you made was because that's the style that comes from PMI.  But there's no reason to follow that style.  I like the idea of changing it to something more generic.